calyx-opt/src/passes/compile_sync_without_sync_reg.rs

```diff
@@ -0,0 +1,211 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::{self as ir, GetAttributes, RRC};
+use calyx_ir::{build_assignments, guard, structure};
+use calyx_utils::{CalyxResult, Error};
+use std::collections::HashMap;
+
+#[derive(Default)]
+/// Compiles @sync without use of std_sync_reg
+/// Upon encountering @sync, it first instantiates a std_reg(1) for each thread(`bar`)
+/// and a std_reg(1) for each barrier (`s`)
+/// It then continuously assigns the value of (`s.in`) to 1'd1 guarded by the
+/// expression that all values of `bar` for threads under the barrier are
+/// set to 1'd1
+/// Then it replaces the @sync control operator with
+/// seq {
+///     barrier;
+///     clear;
```

Comment:
Does each thread generate a `clear` group or is it just the first thread?

Reply:
Each thread generates a `clear` group

Reply:
So which thread is responsible for clearing `s` (the shared register across threads)

Reply:
Sorry, my mistake. `s` a `std_wire.` Since Calyx assigns all undefined values to 0 this should be fine?

Reply:
Ah got it! Worth clarifying in the comment